srcts/src/components/card.ts

```diff
@@ -0,0 +1,277 @@
+import type { Tooltip as TooltipType } from "bootstrap";
+import { getAllFocusableChildren } from "./_utils";
+import { ShinyResizeObserver } from "./_shinyResizeObserver";
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+const Tooltip = (
+  window.bootstrap ? window.bootstrap.Tooltip : class {}
+) as typeof TooltipType;
+
+interface CardOverlay {
+  container: HTMLDivElement;
+  anchor: HTMLAnchorElement;
+}
+
+class Card {
+  private container: HTMLElement;
+  private overlay: CardOverlay;
+  private lastFocusInterior: HTMLElement | undefined;
+  private prevFocusExterior: HTMLElement | undefined;
+
+  private static attr = {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    ATTR_INIT: "data-bslib-card-init",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    CLASS_CARD: "bslib-card",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    CLASS_FULL_SCREEN: "bslib-full-screen",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    CLASS_HAS_FULL_SCREEN: "bslib-has-full-screen",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    CLASS_FULL_SCREEN_ENTER: "bslib-full-screen-enter",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    CLASS_FULL_SCREEN_EXIT: "bslib-full-screen-exit",
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    ID_FULL_SCREEN_OVERLAY: "bslib-full-screen-overlay",
+  };
+
+  /**
+   * A Shiny-specific resize observer that ensures Shiny outputs in the main
+   * content areas of the sidebar resize appropriately.
+   * @private
+   * @type {ShinyResizeObserver}
+   * @static
+   */
+  private static shinyResizeObserver = new ShinyResizeObserver();
+
+  constructor(el: HTMLElement) {
+    // remove initialization attribute and script
+    el.removeAttribute(Card.attr.ATTR_INIT);
+    el.querySelector<HTMLScriptElement>(
+      `script[${Card.attr.ATTR_INIT}]`
+    )?.remove();
+
+    this.container = el;
+    Card.instanceMap.set(el, this);
+
+    // Let Shiny know to trigger resize when the card size changes
+    // TODO: shiny could/should do this itself (rstudio/shiny#3682)
+    Card.shinyResizeObserver.observe(this.container);
+
+    this._addEventListeners();
+    this._enableTooltips();
+    this.overlay = this._createOverlay();
+  }
+
+  enterFullScreen(event?: Event): void {
+    if (event) event.preventDefault();
+
+    this.overlay.container.addEventListener("click", () =>
+      this.exitFullScreen()
+    );
+    document.addEventListener(
+      "keyup",
+      (ev) => this._exitFullScreenOnEscape(ev),
+      false
+    );
+
+    const focusableElements = getAllFocusableChildren(this.container);
```

Comment:
I suppose technically, we should be re-computing `lastFocusInterior` if the card gets mutated while full-screen (i.e., we need a `MutationObserver()`?)

Reply:
Good point that the interior contents might change. I think we can make it work without `MutationObserver` if we're careful about where we place the event listener, but we'd need to recompute last focusable element on demand.